### DISCUSSION TITLE
ERROR in camera_pi

### DISCUSSION THREAD

**Discussion by jackqu0** [signag/raspi-cam-srv#11] [Q&A]
After struggling with motioneye and other streaming apps for the raspberry pi I was overjoyed at discovering raspiCamSrv - something that just ran when installed. However, I have found that I am getting errors when changing the default configuration:
`Could not import SensorConfiguration from picamera2.configuration. Bypassing sensor configuration`
This appears on the command line from which the flask command is run.

I get the same error on multiple camera's on both a RPi 3B+ and RPi Zero 2W. Confusingly, in the app web config page the camera properties are shown as I would expect.

I see that this is listed as a known error on the Troubleshooting page but just states that "This message may occur when running on Bullseye systems", which is true as both pi's are running bullseye (I don't think I can install bookworm on them).

Any ideas on how to fix and stop it crashing when changing the config?

**Comment by signag** [comment]
Thanks for the feedback.    
I think I wanted to write a little bit more about that "error" but then forgot it.   
It is [updated](https://github.com/signag/raspi-cam-srv/blob/main/docs/Troubelshooting.md) now.

In short: Just ignore it. It has no impact on functionality.

*SensorConfiguration* is a class in Picamera2 which is referenced in the CameraConfiguration (see [Picamera2 Manual](https://datasheets.raspberrypi.com/camera/picamera2-manual.pdf) chapter 4.3)     
It includes information on the output size and bit depth of a stream.

In Bullseye systems, this is missing. Probably, this had been introduced in Bookworm deployments of Picamera2 and not been ported to Bullseye by the Picamera2 team.     
Since there is some rumor that they will make Bookworm available also on those systems which currently run Bullseye, I did not want to go to the Picamera2 team to ask for an update of the Picamera2 Bullseye deployment.

Currently, raspiCamSrv does not require the *SensorConfiguration* but it is included in the data model because Picamera2 uses it.    
Therefore, I put out an error when the class cannot be imported.

**Comment by jackqu0** [comment]
Thanks for a quick response.

I have subsequently discovered that Bookwoom can be installed on the Zero2W. Just trying to set up now. Hopefully this will help with some of the issues I have been seeing.

**Comment by signag** [comment]
Where did you find that?    
The latest imager from [https://www.raspberrypi.com/software/](https://www.raspberrypi.com/software/) is version 1.8.5 and still offers just Bullseye.

**Reply by jackqu0** [reply]
If you go to the Raspberry Pi website/Software the Bookworm image is available and states it supports "All Raspberry Pi models".

**Comment by signag** [comment]
OK, I'll also try the version from [https://www.raspberrypi.com/software/operating-systems/](https://www.raspberrypi.com/software/operating-systems/) on a Pi Zero 2W.     

Which configuration changes did you make which made raspiCamSrv crash?    
I just tried Transform and Sensor Mode on a Pi Zero, which were OK.    
If you try with the highest possible resolutions, you get an error message at the bottom of the screen:    
![image](https://github.com/signag/raspi-cam-srv/assets/75881984/576dff39-f62b-4aa5-88b4-b3946b306398)     
which reminds you to use a lower resolution.     

Such errors come from background threads of Picamera2, such as encoders and not from raspiCamSrv.

**Reply by jackqu0** [reply]
Yes, I get that "Error in camera" - and then the video stream hangs and I have to restart flask. The camera says it supports these higher resolutions so I am confused why they are not properly supported (I realise that this is probably beyond raspiCamSrv!)

I have noted that I also get a warning
[0:12:05.627905135] [1190]  WARN RPiSdn sdn.cpp:39 Using legacy SDN tuning - please consider moving SDN inside rpi.denoise

**Reply by jackqu0** [reply]
BTW, I note that the flask process is running at ~70% utilisation. Is this expected? I am running the "main" stream in Senor mode 2.

FYI, I've also just used the Bookworm lite version.

**Comment by signag** [comment]
My understanding why the high resolutions do not work on a Pi Zero (2W) is that there is not enough memory.    
I also reduce the number of buffers to 2 on Pi Zero whereas this is 6 on a Pi 5.    
The error messages coming from Picamera2, libcamera or the encoder are sometimes not really helpful.    
"No such process" only seems to indicate that the encoder process terminated unexpectedly.     
Within raspiCamSrv you normally do not need to restart the server. Selecting another configuration should be sufficient.    
I would recommend to start from the bottom rather from the top.

The Picamera2 manual documents several restrictions for Pi versions < 5. However, I did not see any related to resolutions.

The WARN RPiSdn warning comes from libcamera (see also Issue #5.

**Reply by jackqu0** [reply]
Thanks for your help and explanations. I've now got I running with autostart and I'll test overnight before deploying into my greenhouse :-)
